pkg/ccl/backupccl/restore_old_versions_test.go

```diff
@@ -730,3 +744,44 @@ func TestRestoreWithDroppedSchemaCorruption(t *testing.T) {
 	}
 	require.Falsef(t, hasSameNameSchema(dbName), "corrupted descriptor exists")
 }
+
+// restorePublicSchemaRemap tests that if we're restoring a database from
+// an older version where the database has a synthetic public schema, a real
+// descriptor backed public schema is created and the tables in the schema
+// are correctly mapped to the new public schema.
+func restorePublicSchemaRemap(exportDir string) func(t *testing.T) {
+	return func(t *testing.T) {
+		const numAccounts = 1000
+		_, _, _, tmpDir, cleanupFn := BackupRestoreTestSetup(t, MultiNode, numAccounts, InitManualReplication)
+		defer cleanupFn()
+
+		_, _, sqlDB, cleanup := backupRestoreTestSetupEmpty(t, singleNode, tmpDir,
+			InitManualReplication, base.TestClusterArgs{})
+		defer cleanup()
+		err := os.Symlink(exportDir, filepath.Join(tmpDir, "foo"))
+		require.NoError(t, err)
+
+		sqlDB.Exec(t, fmt.Sprintf("RESTORE DATABASE %s FROM '%s'", "d", LocalFoo))
+
+		var restoredDBID, publicSchemaID, rowCount int
+		row := sqlDB.QueryRow(t, `SELECT id FROM system.namespace WHERE name='d' AND "parentID"=0`)
+		row.Scan(&restoredDBID)
+		row = sqlDB.QueryRow(t, fmt.Sprintf(`SELECT id FROM system.namespace WHERE name='public' AND "parentID"=%d`, restoredDBID))
+		row.Scan(&publicSchemaID)
+
+		if publicSchemaID == keys.PublicSchemaID {
+			t.Fatalf("expected public schema id to not be %d", keys.PublicSchemaID)
+		}
+
+		row = sqlDB.QueryRow(t,
```

Comment:
QueryRow will error out if it returns more than one row. We can do something like:

row = QueryRow(SELECT * FROM ....)
require.NotNil(row)

Reply:
Done